### PR TITLE
feat: use "serde_json" for "BlobType" string conversion

### LINKED ISSUE #2358: Use `serde_json` in the `Display`/`FromStr` implementations for `BlobType`
`serde_json` https://serde.rs/json.html#structs-and-enums-in-json seems to actually generate the same thing we're doing by hand, so let's use that instead to prevent issues if people forget to modify these when adding new `BlobType`s

### PR DESCRIPTION
**Pull request by duguorong009**
## Motivation
Resolve #2358 

## Proposal
Use the `serde_json::to_string()/from_str()` for `BlobType` string conversion.

## Test Plan
CI

### CODE REVIEW

File: linera-base/src/identifiers.rs [hunk 1]
```diff
@@ -167,27 +167,6 @@ pub enum BlobType {
...
-    type Err = anyhow::Error;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s {
-            "Data" => Ok(BlobType::Data),
-            _ => Err(anyhow!("Invalid blob type: {}", s)),
-        }
-    }
-}
-
```
Comment (andresilva91) [inline linera-base/src/identifiers.rs:190]: I would rather keep these implementations and use `serde_json` in here instead, that way we can also change a lot less code

File: CLI.md [hunk 2]
```diff
@@ -222,7 +222,9 @@ Open (i.e. activate) a new multi-owner chain deriving the UID from an existing o
 * `--from <CHAIN_ID>` — Chain ID (must be one of our chains)
 * `--super-owner-public-keys <SUPER_OWNER_PUBLIC_KEYS>` — Public keys of the new super owners
 * `--owner-public-keys <OWNER_PUBLIC_KEYS>` — Public keys of the new regular owners
-* `--owner-weights <OWNER_WEIGHTS>` — Weights for the new owners
+* `--owner-weights <OWNER_WEIGHTS>` — Weights for the new owners.
+
+   If they are specified there must be exactly one weight for each owner. If no weights are given, every owner will have weight 100.
```
Comment (andresilva91) [inline CLI.md:227]: Could you move these documentation changes to a separate PR?

**Review by andresilva91 — CHANGES_REQUESTED** [review]
Thank you for the PR! I have a couple comments

File: linera-base/src/identifiers.rs (hunk 1 above)
Comment (duguorong009) [reply linera-base/src/identifiers.rs]: Done https://github.com/linera-io/linera-protocol/pull/2364/commits/81bc43cf5cb39ac863f0c840949f8cd42ebc00b4

File: CLI.md (hunk 2 above)
Comment (duguorong009) [reply CLI.md:227]: Hmm, I didn't wanna add this MD file, too.
But, the CI complain made me rebuild the `CLI.md` file. 
https://github.com/linera-io/linera-protocol/actions/runs/10306509270/job/28529761766

Should I roll back this change regardless of the CI complaints?

File: linera-base/src/identifiers.rs [hunk 3]
```diff
@@ -169,22 +169,15 @@ pub enum BlobType {
...
 
 impl FromStr for BlobType {
     type Err = anyhow::Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s {
-            "Data" => Ok(BlobType::Data),
-            _ => Err(anyhow!("Invalid blob type: {}", s)),
-        }
+        serde_json::from_str(s).map_err(|_| anyhow!("Invalid BlobType: {}", s))
```
Comment (afck) [inline linera-base/src/identifiers.rs:180]:
```suggestion
        serde_json::from_str(s).with_context(|| format!("Invalid BlobType: {}", s))
```

That way the original error is included in the `anyhow` error as well.

File: examples/Cargo.lock [hunk 4]
```diff
@@ -3321,6 +3321,7 @@ dependencies = [
  "serde",
  "serde-name",
  "serde_bytes",
+ "serde_json",
```
Comment (afck) [inline examples/Cargo.lock:3324]: That's probably not needed?

File: CLI.md (hunk 2 above)
Comment (afck) [reply CLI.md:227]: It's probably because you upgraded `clap` (and lots of other crates).

Let's revert the `Cargo.toml` changes (other than what's necessary to add serde) and this one, since they are not related to this change.

File: examples/Cargo.lock (hunk 4 above)
Comment (duguorong009) [reply examples/Cargo.lock:3324]: That is needed for use of `serde_json::*` in code.

File: linera-base/src/identifiers.rs (hunk 3 above)
Comment (duguorong009) [reply linera-base/src/identifiers.rs]: Thx! Done https://github.com/linera-io/linera-protocol/pull/2364/commits/cfd2bd1e13b91677abb6b162831b39f0a3d35aa6

**Comment by duguorong009** [comment]
@afck 
Thx for the review!

> It's probably because you upgraded clap (and lots of other crates).
> 
> Let's revert the Cargo.toml changes (other than what's necessary to add serde) and this one, since they are not related to this change.  

Yes, I think the reason of difference in CLI.md is attributed to `clap` & `clap_markdown` crate.
Will reverse their versions.

Fyi, I run the `cargo update` for updating the crates.
That's why there is huge diff in `Cargo.lock`.
The reason is that current `Cargo.lock` fails to build the project in my machine(Mac-intel).
Maybe, can we merge this new `Cargo.lock`?

**Comment by afck** [comment]
Our policy is to not do `cargo update` in unrelated PRs. We'd rather do a dedicated, separate PR with _only_ the `cargo update` changes.

So for this PR, I'd prefer to revert all the unrelated updates.

**Comment by duguorong009** [comment]
> Our policy is to not do `cargo update` in unrelated PRs. We'd rather do a dedicated, separate PR with _only_ the `cargo update` changes.
> 
> So for this PR, I'd prefer to revert all the unrelated updates.

okay, I revert the `Cargo.lock`.

I believe you don't mind that I create a separate PR for `Cargo update`. right?

File: linera-base/src/identifiers.rs [hunk 5]
```diff
@@ -169,22 +169,15 @@ pub enum BlobType {
 
 impl Display for BlobType {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            BlobType::Data => write!(f, "Data")?,
-        };
-
-        Ok(())
+        write!(f, "{}", serde_json::to_string(self).unwrap())
```
Comment (afck) [inline linera-base/src/identifiers.rs:172]: I think there's a dedicated `to_writer` function for this. Also, let's avoid `unwrap`ping:

```suggestion
        serde_json::to_writer(f, self).map_err(|_| fmt::Error)
```

**Review by afck — COMMENTED** [review]
Thanks, looks great! :+1: 
Just one more nit-pick from my side. And let's wait for @andresilva91 to approve, too.

I'm also in favor of a separate PR for the `clap` version update: Adding the longer form of the documentation to `CLI.md` makes sense to me!

File: linera-base/src/identifiers.rs (hunk 5 above)
Comment (duguorong009) [reply linera-base/src/identifiers.rs:172]: Afaik, the `to_writer()` needs the `f` implement the `std::io::Write` trait. 
But, `std::fmt::Formatter` does not meet the requirements.
I believe we should stick to `to_string()`. Then, maybe, we cannot avoid `unwrap`.

Comment (afck) [reply linera-base/src/identifiers.rs:172]: But `Formatter` _does_ imply `Write`, doesn't it?
https://doc.rust-lang.org/nightly/std/fmt/struct.Formatter.html#impl-Write-for-Formatter%3C'_%3E

And we can always avoid `unwrap` using `.map_err(|_| fmt::Error)`.

Comment (duguorong009) [reply linera-base/src/identifiers.rs:172]: Is that `Write` equal to `std::fmt::Write`? not the `std::io::Write`.
`serde_json::to_writer` needs the writer to implement the `std::io::Write`, targeting the I/O stream.
https://docs.rs/serde_json/latest/serde_json/fn.to_writer.html

Also, if we really wanna avoid `unwrap`, how about this?
```
impl Display for BlobType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match serde_json::to_string(self) {
            Ok(s) => write!(f, "{}", s),
            Err(_) => return Err(fmt::Error),
        }
    }
}
```

Comment (afck) [reply linera-base/src/identifiers.rs:172]: Ah, I see! Right, I guess we can't use `to_writer` then.

Sure, I'm fine with the `match`, too, if you prefer that. :+1:

**Review by afck — APPROVED** [review]
Thanks! :+1:

**Review by andresilva91 — APPROVED** [review]
Looks great! Thank you
[shipit](https://app.graphite.dev//settings/meme-library?org=linera-io)

**Comment by duguorong009** [comment]
I believe you can merge the PR, not me! 😅 @andresilva91 @afck
